### DISCUSSION TITLE
Why doesn't c_str() passed to popen() work as intended in Natve Messaging host?

### DISCUSSION THREAD

**Discussion by guest271314** [nlohmann/json#3399] [Q&A] [answered]
I am currently using `<nlohmann/json.hpp>` from nlohmann-json3-dev (following the example at https://discourse.mozilla.org/t/webextension-with-native-messaging-c-app-side/30821) to process Native Messaging input and output here https://github.com/guest271314/captureSystemAudio/blob/master/native_messaging/capture_system_audio/capture_system_audio.cpp, which works as intended. 

The reason I ask this question is I am not sure if I really need the entirety of json.hpp to achieve the requirement. In the working code this appears to be the lines which allow `popen()` to execute the parameter passed https://github.com/guest271314/captureSystemAudio/blob/543984c26f210d747d8eb843eb3429926c414c91/native_messaging/capture_system_audio/capture_system_audio.cpp#L26-L27

```
string m(message, message + sizeof message / sizeof message[0]);
return json::parse(m);
```

The Native Messaging protocol is described in pertinent part here https://developer.chrome.com/docs/apps/nativeMessaging/#native-messaging-host-protocol:

> Chrome starts each native messaging host in a separate process and communicates with it using standard input (stdin) and standard output (stdout). The same format is used to send messages in both directions: each message is serialized using JSON, UTF-8 encoded and is preceded with 32-bit message length in native byte order.

I asked a similar question here https://www.reddit.com/r/cpp_questions/comments/sxv9cv/string_to_const_char_for_popen/ before I found the example on discourse.mozilla.org.

What I do not yet understand is why `popen()` does not execute the result of `c_str()` from Native Messaging client input.

When I send the input back to Native Messaging client I get expected result; in this case `parec -d @DEFAULT_MONITOR` .

However, `popen()` will only execute when I hard code the string.

Attached is a screenshot of what is described above. The JavaScript arrays in the top of the screenshot are sent from Native Messaging host when 

```
FILE* pipe = popen("parec -d @DEFAULT_MONITOR", "r");
``` 

is hardcoded in the .cpp file - not when 

```
auto command = inStr.c_str();
sendMessage(command);
FILE* pipe = popen(command, "r");
```
is used.

Why?

```
#include <iostream>
#include <cstdio>
#include <string>
#include <cstring>
#include <vector>
#include <iomanip>
#include <sstream>

using namespace std;

void sendMessage(string message) {
  auto* data = message.data();
  auto size = uint32_t(message.size());
  cout.write(reinterpret_cast<char*>(&size), 4);
  cout.write(data, size);
  cout.flush();
}

int main() {
  while (true) {
    // https://github.com/Mandar-Shinde/Google-Chrome-Native-Messaging/blob/master/Client-C%2B%2B/Worker.cpp
    char len[4];
    cin.read(len, 4);
    unsigned int ilen = *reinterpret_cast<unsigned int*>(len);
    if (ilen < 1024 && ilen >= 1) {
      char* inMsg = new char[ilen];
      memset(inMsg, 0, ilen + 1);
      cin.read(inMsg, ilen);
      string inStr(inMsg);
      delete[] inMsg;
      auto command = inStr.c_str();
      sendMessage(command);
      FILE* pipe = popen(command, "r");
      while (true) {
        unsigned char buffer[1024];
        int count = fread(buffer, 1, sizeof buffer, pipe);
        if (count == 0 || count == -1) {
          break;
        }
        stringstream ss;
        ss << "[";
        for (int i = 0; i < count; i++) {
          ss << (int)buffer[i];
          if (i < count - 1) {
            ss << ",";
          }
        }
        ss << "]";
        sendMessage(ss.str());
      }
    }
  }
}
```

![Screenshot_2022-03-20_09-57-57](https://user-images.githubusercontent.com/4174848/159174015-10286d67-1257-465a-95eb-95203477052d.png)

**Comment by gregmarr** [comment]
I don't see how this question has anything to do with the json library itself.

**Reply by guest271314** [reply]
The code in https://discourse.mozilla.org/t/webextension-with-native-messaging-c-app-side/30821 uses this library. I am trying to determine precisely what this library does to achieve the expected results that C++ string does not do.

**Reply by nlohmann** [reply]
In the example you linked, it really seems there is no added value by using `json` over a `std::string`.

**Reply by guest271314** [reply]
When parsing input I notice that characters are missing/stripped from input when `json` is _not_ used, which results in `popen()` not being executed when parameters are passed as a command. 

I am trying to isolate precisely which parts of `json` do exactly what to preserve original input string.

**Reply by guest271314** [reply]
What I don't understand is why space characters are making it to `popen()`. If I `postMessge("ls")` I get the output back. When I `postMessage("ls -l")` I do not.

**Comment by guest271314** [comment, marked as answer]
@nlohmann I found a way to parse the input so that it can be used with `popen()`: exclude beginning and ending double quotation marks. Now I can use `stringstream` instead of using your json.hpp file - which greatly helped guiding me towards achieving the requirement:

```
// C++ Native Messaging host
// https://browserext.github.io/native-messaging/
// https://developer.chrome.com/docs/apps/nativeMessaging/
// https://discourse.mozilla.org/t/webextension-with-native-messaging-c-app-side/30821
#include <iostream>
using namespace std;
#include <sstream>

void sendMessage(string message) {
  auto* data = message.data();
  auto size = uint32_t(message.size());
  char* length = reinterpret_cast<char*>(&size);
  fwrite(length, 4, sizeof(char), stdout);
  fwrite(message.c_str(), message.length(), sizeof(char), stdout);
  fflush(stdout);
}

string getMessage() {
  char length[4];
  fread(length, 4, sizeof(char), stdin);
  uint32_t len = *reinterpret_cast<uint32_t*>(length);
  if (!len) {
    exit(EXIT_SUCCESS);
  }
  char message[len];
  fread(message, len, sizeof(char), stdin);
  string content(message, message + sizeof message / sizeof message[0]);
  return content;
}

int main() {
  while (true) {
    auto message = getMessage();
    const char* command = message.data();
    stringstream input;
    // Exclude double quotation marks from beginning and end of string
    for (int j = 1; j < message.length() - 1; j++) {
      input << message[j];
    }
    FILE* pipe = popen(input.str().c_str(), "r");
    while (true) {
      unsigned char buffer[1024];
      int count = fread(buffer, 1, sizeof buffer, pipe);
      if (count == 0 || count == -1) {
        break;
      }
      stringstream output;
      output << "[";
      for (int i = 0; i < count; i++) {
        output << (int)buffer[i];
        if (i < count - 1) {
          output << ",";
        }
      }
      output << "]";
      sendMessage(output.str());
    }
  }
}
```
